src/Control/SessionHandler/DatabaseSessionHandler.php

```diff
@@ -0,0 +1,218 @@
+<?php
+
+namespace SilverStripe\Control\SessionHandler;
+
+use SensitiveParameter;
+use SilverStripe\Core\Config\Configurable;
+use SilverStripe\Core\Convert;
+use SilverStripe\ORM\Connect\Query;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\DB;
+use SilverStripe\ORM\FieldType\DBDatetime;
+use SilverStripe\ORM\Queries\SQLDelete;
+use SilverStripe\ORM\Queries\SQLInsert;
+use SilverStripe\ORM\Queries\SQLSelect;
+use SilverStripe\ORM\Queries\SQLUpdate;
+
+/**
+ * Session save handler that stores session data in the database.
+ */
+class DatabaseSessionHandler extends AbstractSessionHandler
+{
+    use Configurable;
+
+    private static string $table_name = '_sessions';
+
+    public function open(string $path, string $name): bool
+    {
+        // No action is required to open the session.
+        return true;
+    }
+
+    public function close(): bool
+    {
+        // No action is required to close the session.
+        return true;
+    }
+
+    /**
+     * @inheritDoc
+     * Clears the cache entry that represents this session ID.
+     */
+    public function destroy(#[SensitiveParameter] string $id): bool
+    {
+        if (!$this->isDatabaseReady()) {
+            return false;
+        }
+
+        SQLDelete::create(
+            Convert::symbol2sql(static::config()->get('table_name')),
+            [Convert::symbol2sql('ID') => $id]
+        )->execute();
+        return true;
+    }
+
+    /**
+     * @inheritDoc
+     * Clears all session cache which have a last modified datetime older than the session max lifetime.
+     * Note that we use our own calculated session lifetime rather than the passed in lifetime which doesn't
+     * take Silverstripe CMS configuration values into account.
+     */
+    public function gc(int $max_lifetime): int|false
+    {
+        if (!$this->isDatabaseReady()) {
+            return false;
+        }
+
+        SQLDelete::create(
+            Convert::symbol2sql(static::config()->get('table_name')),
+            [Convert::symbol2sql('Expiry') . ' < ?' => DBDatetime::now()->getTimestamp()]
+        )->execute();
+        return DB::affected_rows();
+    }
+
+    /**
+     * @inheritDoc
+     * Returns data of a pre-existing session, or an empty string for a new session.
+     */
+    public function read(#[SensitiveParameter] string $id): string|false
+    {
+        if (!$this->isDatabaseReady()) {
+            return false;
```

Comment:
I'm wondering if we'd be better off just throwing an exception if isDatabaseRead() returns false. What happens in all these methods like open() / gc() / read() etc if `false` is returned?  Is there some risk of things getting out of sync and acting weird if we don't have a hard failure?

Reply:
The API for this method and other methods that can return false is defined by PHP itself - if the intention was for hard exceptions to be thrown, PHP would have added a `@throws` to the PHPdoc.

I'll add some logging though, that seems sensible.